### PR TITLE
flesh out saltify driver

### PR DESCRIPTION
**Pull request by vernondcole**
Note: this is a re-submission of #41940 which had git merge errors.

### What does this PR do?

Extends the saltify driver so that it will support the `-d`, `-Q`, `-F`, `-a reboot`, and `--list-images` commands. Inappropriate commands (such as `--list-locations`) return an empty result.

In order to implement the above, a new feature is added to the body of salt-cloud, so that
for all new minions, a `salt-cloud` grain is added which identifies the driver, the provider, 
and the profile for the minion. This is controlled by the option `enable_cloud_grains`
which defaults as `True`. 

Also, to fix the bug where a new minion (created by `salt-cloud -p xxx yyyy`) cannot contact
its salt master because it had pre-existing configuration for a different master, a new feature
is added to force overwrite of the configuration.  This is controlled by the option
`force_minion_config` which defaults as `False`.

The new saltify `--destroy` implementation can partially clean up what the deployment created
by removing the minion configuration (controlled by the `remove_config_on_destroy` option, 
which defaults as `True`) and shut down the minon (controlled by the `shutdown_on_destroy`
option which defaults as `False`).

Note: salt-api must be enabled on the salt master for the new commands to work. A brief example of enabling it is included in the updated documentation.

### What issues does this PR fix or reference?

#40962 

### Previous Behavior

```
bvm $ sudo salt-cloud -p hpltx2 admindb
[...clip...]
Connection to 10.100.9.41 closed.
 *  INFO: Salt installed!
[...clip...]
bvm $ sudo salt-call -d admindb
No machines were found to be destroyed 
```

### New Behavior

```
bvm $ sudo salt-call -d admindb
Deleted admindb
```
Also, the `-Q`, `-F`, `-a reboot`, and `--list-images` commands work.

### Tests written?

No. The existing tests of saltify have not been extended.

### CODE REVIEW

**Review by techhat — APPROVED** [review]

**Review by gtmanfred — APPROVED** [review]

**Comment by vernondcole** [comment]
Sorry. Inadvertently deleted the PR while trying to clean up the bad version.

File: salt/utils/cloud.py
Comment (vernondcole) [inline salt/utils/cloud.py]: My patch for salt-bootstrap #1105 has been merged.  We should pull this out before merging here.
I will prepare the update as soon as I can.

**Review by vernondcole — COMMENTED** [review]
Pulled the workaround. My PR # 1105 for salt-bootstrap has been merged, so the work-around for it is no longer needed. (I assume that salt-bootstrap has a shorter release cycle that salt.) I pulled it out removing the ugly extra invocation of the bootstrap code.

**Comment by vernondcole** [comment]
Guys:  I am using this code as the basis for the Vagrant driver I am working on now & have merged this with my branch for that, so this ought to be merged before I can start a PR for Vagrant. Are we waiting for the Nitrogen release before this can be merged? Someone with write approval must push the button, I cannot.  Do I need to rebase again?

**Comment by vernondcole** [comment]
@gtmanfred @techhat  These failures all seem to be Python3 issues in modules I have not touched.  I fixed my one PY3 bug (importing the wrong version of ipaddress) this morning, and rebased my branch.  Can this be merged now?

**Comment by rallytime** [comment]
@vernondcole The python3 errors should all be cleared up now on develop. I've updated your branch to the head of develop and restarted the tests. Pending the test run, this is ready to go from your perspective, yes?

**Comment by vernondcole** [comment]
I was double-checking the documentation this morning and noticed that it
could use a couple of tweaks. I am setting up a different test apparatus to
make sure what I wrote actually works in a general case..  If it needs
repair, I can submit that as a follow-up PR, or I could just sneak it in
with the Pull Request I plan to submit in two weeks for a Vagrant
salt-cloud driver.  As for the actual software, yes, it is ready to go.

But, If you don't mind hanging on for a day or two for a documentation
update, that would be good, too.

**Comment by rallytime** [comment]
@vernondcole No problem at all. :) We can hold on here until you're ready with the docs. Just let us know when you're ready.

**Comment by vernondcole** [comment]
Okay. I streamlined the docs and made several passes at spinning up new VMs and virgin hardware machines to make sure the updated examples actually work.  In the process, I discovered a code change that Git managed to drop from my source (grrrr) and restored it. I also noticed that I was continually having to perform the same manual step (defining a cache directory) to keep my tests from failing.  Looking at that code, it seemed that a try/except block was shorter than intended, so I moved the "try" statement to include missing, as well as unreadable, directories.  

That last one worries me, since it is digging in someone else's ditch. The change makes _my_ code work, but will it break something else?  

The docs are now good to go.

**Comment by vernondcole** [comment]
Documentation is now in good shape.  Latest Jenkins CI run fails, then
says that there are no identified errors... so I am thinking that the
problem is in Jenkins, not my code.

File: doc/topics/cloud/misc.rst
Comment (rallytime) [inline doc/topics/cloud/misc.rst]: This should be `Oxygen` here.

File: salt/cloud/clouds/saltify.py
Comment (rallytime) [inline salt/cloud/clouds/saltify.py]: Maybe turn this example into a bash code-block?

File: salt/cloud/clouds/saltify.py
Comment (rallytime) [inline salt/cloud/clouds/saltify.py]: Can you split this into two lines with the `'''` and `This function...` on their own lines?

File: salt/cloud/clouds/saltify.py
Comment (rallytime) [inline salt/cloud/clouds/saltify.py]: There are several references that look this way in the functions below as well. Adding these docs are great, but providing the actual command examples formatted with `.. code-block:: bash` would be stellar.

File: salt/cloud/clouds/saltify.py
Comment (rallytime) [inline salt/cloud/clouds/saltify.py]: Please split this doc ref into 2 lines as well.

File: salt/cloud/clouds/saltify.py
Comment (rallytime) [inline salt/cloud/clouds/saltify.py]: This version will be `Oxygen`.

File: salt/cloud/clouds/saltify.py
Comment (rallytime) [inline salt/cloud/clouds/saltify.py]: There are a couple of other doc lines that are set-up this way. Can you go through and fix the rest as well?

File: salt/cloud/clouds/saltify.py
Comment (rallytime) [inline salt/cloud/clouds/saltify.py]: All of the versionadded tags that you're adding with this PR should be `Oxygen`. Can you update the remaining ones as well, instead of me commenting on them all? :)

File: salt/utils/cloud.py
Comment (rallytime) [inline salt/utils/cloud.py]: What is the reasoning behind removing this?

File: salt/utils/cloud.py
Comment (rallytime) [inline salt/utils/cloud.py]: Same question here: why is this being removed?

**Review by rallytime — DISMISSED** [review]
Hi @vernondcole - thank you for all of your work here with this! I had a couple of comments and questions. If you could address those, that would be excellent.

**Comment by rallytime** [comment]
@vernondcole In regards to the Jenkins results, you're correct in that the failure there isn't related to your changes here.

File: salt/cloud/clouds/saltify.py
Comment (vernondcole) [inline salt/cloud/clouds/saltify.py]: I would love to put a CLI example with this one, but I have no idea how to call this function. There are no examples in any of the other drivers. Since I don't know how to use it, I can't test it.  Any body care to supply a working example?

File: salt/utils/cloud.py
Comment (vernondcole) [inline salt/utils/cloud.py]: Same answer: it is dead code.

File: salt/utils/cloud.py
Comment (vernondcole) [inline salt/utils/cloud.py]: As it happens, this parameter was  stepped on in the implementation, so anything passed would be useless.  Under the principle of least surprise, a parameter which looks like it should work, but does not, seems to be a bad idea.  The value of `deploy_command` is built from a config.get call at about line 1246.  (I would not have noticed this -- PyCharm actually caught the error.)

**Comment by vernondcole** [comment]
I did a rebase, fixed the merge conflict in client/__init__.py, and did a push --force as documented in https://docs.saltstack.com/en/latest/topics/development/contributing.html.  I hope that worked right.  Have I ever mentioned that I really hate git?

File: salt/cloud/clouds/saltify.py
Comment (vernondcole) [inline salt/cloud/clouds/saltify.py]: Sorry -- that comment was supposed to apply to list_nodes_select

**Comment by gtmanfred** [comment]
@vernondcole is there a reason that you keep pushing a merge commit instead of rebasing?

This needs to be rebased with your changes on top, not merged into the middle.

You also don't need to keep pushing, you could have just clicked the `Update Branch` button on github.

**Comment by vernondcole** [comment]
@gtmanfred  I am trying to follow the flow documented in [https://docs.saltstack.com/en/latest/topics/development/tutorial.html](url). The only command I have typed in the last week is "rebase" -- but what I get gets worse and worse.  I would be happy to do exactly what you request, but I don't know how.  (Have I mentioned my real dislike for git?)  Also -- my `Udate branch` button is greyed out, so that's not an option for me.  Tell me what to type, and I will type it.

**Comment by vernondcole** [comment]
@rallytime Were my answers to your questions insufficient?  I seem to still be blocked.  Is there a problem with removing the dead code?  I did search for other calls to the function and found none.

**Comment by gtmanfred** [comment]
You will need to do use `git rebase -i` and remove the extra commits.

When updating the branch, you can do `git pull --rebase upstream develop`

Also, you do not need to update it then, we can push the update Branch, and
when it gets to reviewing, we can update it.

File: salt/utils/cloud.py
Comment (rallytime) [inline salt/utils/cloud.py]: Ah, you're right. Thanks for pointing that out.

**Review by rallytime — COMMENTED** [review]
The remaining question I had has been resolved. I'd like to look over this one more time once the rebasing issues are resolved.

**Comment by vernondcole** [comment]
So far, so good.  I have run an interactive rebase and believe that I have the correct thing on my workstation. This may be where I go astray.

>$ git status
On branch enhancement/saltify_d
Your branch and 'origin/enhancement/saltify_d' have diverged,
and have 1 and 29 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
nothing to commit, working tree clean

I have been following git's advice and doing the pull.  Am I supposed to do `git push --force` here?

Or what?

**Comment by gtmanfred** [comment]
Yeah, you will need to do a force push, because you are writing changes to
the order of commits to the branch.

**Comment by vernondcole** [comment]
Looks to me like the failing test is a Jenkins problem... inside the console log I see:

> 19:09:19 ================================================================================================================================================================
19:09:19 OK (total=6989, skipped=574, passed=6415, failures=0, errors=0) 
19:09:19 ====================================================================  Overall Tests Report  ====================================================================

**Comment by gtmanfred** [comment]
yup, the default does that from time to time, this should be good when
nicole gets back to it to merge it :+1:

File: salt/utils/cloud.py
Comment (rallytime) [inline salt/utils/cloud.py]: Is there a reason you're changing the name of this function? It is unclear to me why these changes are necessary and we can't move the names of functions like this without putting them on a deprecation path.

**Review by rallytime — COMMENTED** [review]
Hi @vernondcole - 
I have a follow up question on one of the changes in `salt.utils.cloud.py`. There are quite a few changes happening in this file. These functions are used widely throughout salt-cloud, so I just want to make sure we're treading carefully here. :)

File: salt/utils/cloud.py
Comment (vernondcole) [inline salt/utils/cloud.py]: Good catch!  The former function, cache_nodes_id, was never called anywhere in salt.  It appears that I replaced it with one that I was planning to use, get_cache_node, but abandoned when that design did not work out, and then forgot about.  So it is never called either.  So both versions are dead code.
  What would be the best path forward?
. keep both dead versions?
. keep my new dead version?
. keep the original dead version?
. kill both?

Comment (vernondcole) [inline salt/utils/cloud.py]: Oh ... one other thing ... the original function always returns "None".

**Comment by vernondcole** [comment]
@rallytime I pulled my new unused code and added a depreciation warning for the existing dead function.

**Comment by vernondcole** [comment]
Arrrrgh!  Did git just eat all of my commits except for the two in conflict???
Looking for my lost code now.

**Comment by vernondcole** [comment]
Restored the missing changes. This should be ready now. Again.

File: salt/utils/cloud.py
Comment (rallytime) [inline salt/utils/cloud.py]: Can we flesh out this message a little bit more? Maybe just add something like `This function will be removed in Salt Fluorine.` to the end of what you already have. :)

**Review by rallytime — CHANGES_REQUESTED** [review]
@vernondcole This is getting so close! I just had one question and another comment. After those two things are resolved, we can get this in. Thank you very much for all of your work here.

**Comment by rallytime** [comment]
Oh, and there are also a couple of minor lint errors here as well taht need to be fixed: 
https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/13362/violations/file/salt/cloud/clouds/saltify.py/

File: salt/client/__init__.py
Comment (vernondcole) [inline salt/client/__init__.py]: It was causing my tests to fail. Easier to fix the code than add a work-around in the test. Yes, it is pretty much unrelated.

Comment (rallytime) [inline salt/client/__init__.py]: @cachedout What do you think about moving this in the try/except below? I wanted to get your opinion on this. Otherwise this PR looks good to me.

Comment (rallytime) [inline salt/client/__init__.py]: @vernondcole I spoke with @cachedout about this offline. I'd rather see a change like this in a different PR so we can evaluate it as its own change. Can you make a new PR with this change and once that has been evaluated/merged in, it can be included here? Thanks!

**Comment by vernondcole** [comment]
Have moved the client/init code to a different PR as requested, and re-removed the saltify cross-references from loader (again). Sorry for the slow response -- I am on vacation and working from my laptop.

**Review by rallytime — APPROVED** [review]

**Comment by rallytime** [comment]
Thanks @vernondcole!

**Pull request merged**
